### PR TITLE
zh: enable TidbCluster Auto-scaling

### PR DESCRIPTION
**Pull request by Yisaer**
### What is changed, added or deleted? (Required)

Add TidbCluster Autoscaling document

### Which TiDB Operator version(s) do your changes apply to? (Required)

- [x] master (the latest development version, including v1.1 changes for now)

### CODE REVIEW

**Comment by Yisaer** [comment]
@DanielZhangQD  PTAL

File: zh/enable-tidb-cluster-auto-scaling.md [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+        resource:
+          name: "cpu"
+          target:
+            type: "Utilization"
+            averageUtilization: 80
+```
+
+对于 TiDB 组件，你可以通过 `spec.tidb` 来进行配置，目前 tikv 与 tidb 的弹性 api 相同。
+
+在 `TidbClusterAutoScaler` 对象中，`cluster` 属性代表了需要被弹性调度的 Tidb 集群，通过 name 与 namespace 来代表。 由于 `TidbClusterAutoScaler` 组件需要通过指标采集组件抓取相关资源使用情况，我们需要提供对应的指标采集与查询服务给 `TidbClusterAutoScaler`。`monitor` 属性则代表了与之相连的 TidbMonitor 对象。 如果你并不了解 TidbMonitor，可以参考 [通过 TidbMonitor 监控 TiDB 集群](monitor-using-tidbmonitor.md)
```
Comment (ran-huang) [inline zh/enable-tidb-cluster-auto-scaling.md]:
```suggestion
在 `TidbClusterAutoScaler` 对象中，`cluster` 属性代表了需要被弹性调度的 TiDB 集群，通过 name 与 namespace 来代表。 由于 `TidbClusterAutoScaler` 组件需要通过指标采集组件抓取相关资源使用情况，我们需要提供对应的指标采集与查询服务给 `TidbClusterAutoScaler`。`monitor` 属性则代表了与之相连的 TidbMonitor 对象。 如果你并不了解 TidbMonitor，可以参考[通过 TidbMonitor 监控 TiDB 集群](monitor-using-tidbmonitor.md)
```

File: zh/enable-tidb-cluster-auto-scaling.md [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+
```
Comment (ran-huang) [inline zh/enable-tidb-cluster-auto-scaling.md]:
```suggestion
```

File: zh/enable-tidb-cluster-auto-scaling.md [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+
+对于非 `TidbMonitor` 的外部 `Prometheus`, 你可以通过 `spec.metricsUrl` 来填写这个服务的 Host ,从而指定该 TiDB 集群的监控指标采集服务。对于使用 `helm` 部署 TiDB 集群监控的情况，可以通过以下方式来指定 `spec.metricsUrl`。
```
Comment (ran-huang) [inline zh/enable-tidb-cluster-auto-scaling.md]:
```suggestion
对于非 `TidbMonitor` 的外部 `Prometheus`, 你可以通过 `spec.metricsUrl` 来填写这个服务的 Host，从而指定该 TiDB 集群的监控指标采集服务。对于使用 `helm` 部署 TiDB 集群监控的情况，可以通过以下方式来指定 `spec.metricsUrl`。
```

File: zh/enable-tidb-cluster-auto-scaling.md [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+  cluster:
+    name: auto-scaling-demo
+    namespace: default
+  metricsUrl: "http://<release-name>-prometheus.<release-namespace>.svc:9090"
+  ......
+```
+
+`minReplicas` 与 `maxReplicas` 则代表了对于目标组件弹性的上下限。
+
+目前 `TidbClusterAutoScaler` 仅支持基于 CPU 负载的弹性伸缩，CPU 负载的描述性 API 如下所示。`averageUtilization` 则代表了 CPU 负载利用率的 阈值。 当当前 CPU 利用率超过百分之 80 时 则会触发弹性伸。
```
Comment (ran-huang) [inline zh/enable-tidb-cluster-auto-scaling.md]:
```suggestion
目前 `TidbClusterAutoScaler` 仅支持基于 CPU 负载的弹性伸缩，CPU 负载的描述性 API 如下所示。`averageUtilization` 则代表了 CPU 负载利用率的阈值。当当前 CPU 利用率超过百分之 80 时则会触发弹性伸。
```

File: zh/enable-tidb-cluster-auto-scaling.md [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+```shell
+$ kubectl apply -f examples/auto-scale/  -n <namespace>
+tidbclusterautoscaler.pingcap.com/auto-scaling-demo created
+tidbcluster.pingcap.com/auto-scaling-demo created
+tidbmonitor.pingcap.com/auto-scaling-demo created
+```
+
+当 TiDB 集群创建完毕以后，你可以通过 [sysbench](https://www.percona.com/blog/tag/sysbench/) 等数据库压测工具进行压测来体验弹性伸缩功能。
+
+销毁环境
```
Comment (ran-huang) [inline zh/enable-tidb-cluster-auto-scaling.md]:
```suggestion
使用如下命令销毁环境：
```

File: zh/enable-tidb-cluster-auto-scaling.md [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+```shell
+kubectl delete tidbcluster auto-scaling-demo -n <namespace>
+kubectl delete tidbmonitor auto-scaling-demo -n <namespace>
+kubectl delete tidbclusterautoscaler auto-scaling-demo -n <namespace>
+```
+
+## 配置 TidbClusterAutoScaler
+
+对比起无状态的 Web 服务，一个分布式数据库软件对于实例的伸缩往往是非常敏感的。我们需要保证每次弹性伸缩之间存在一定的间隔，从而避免引起频繁的弹性伸缩。
+你可以通过 `spec.tikv.scaleOutThreshold` 和 `spec.tikv.scaleInThreshold` 来配置每两次弹性伸缩之间的时间间隔，对于 tidb 也同样如此。
```
Comment (ran-huang) [inline zh/enable-tidb-cluster-auto-scaling.md]:
```suggestion
你可以通过 `spec.tikv.scaleOutThreshold` 和 `spec.tikv.scaleInThreshold` 来配置每两次弹性伸缩之间的时间间隔，对于 TiDB 也同样如此。
```

File: zh/enable-tidb-cluster-auto-scaling.md [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+    metrics:
+      - type: "Resource"
+        resource:
+          name: "cpu"
+          target:
+            type: "Utilization"
+            averageUtilization: 80
+```
+
+对于 TiDB 组件，你可以通过 `spec.tidb` 来进行配置，目前 tikv 与 tidb 的弹性 api 相同。
```
Comment (DanielZhangQD) [inline zh/enable-tidb-cluster-auto-scaling.md]:
```suggestion
对于 TiDB 组件，你可以通过 `spec.tidb` 来进行配置，目前 TiKV 与 TiDB 的弹性伸缩 API 相同。
```

File: zh/enable-tidb-cluster-auto-scaling.md [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+          name: "cpu"
+          target:
+            type: "Utilization"
+            averageUtilization: 80
+```
+
+
+## 快速上手
+
+我们将通过以下指令快速部署一个 3 PD / 3 TiKV/ 2 TiDB，并带有监控与弹性伸缩能力的 TiDB 集群。
```
Comment (DanielZhangQD) [inline zh/enable-tidb-cluster-auto-scaling.md]:
```suggestion
我们将通过以下指令快速部署一个 3 PD、3 TiKV、2 TiDB，并带有监控与弹性伸缩能力的 TiDB 集群。
```

File: zh/enable-tidb-cluster-auto-scaling.md [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+
+```shell
+$ kubectl apply -f examples/auto-scale/  -n <namespace>
```
Comment (DanielZhangQD) [inline zh/enable-tidb-cluster-auto-scaling.md]: Need to request the users to clone the code first?

File: zh/enable-tidb-cluster-auto-scaling.md [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+```
+
+当 TiDB 集群创建完毕以后，你可以通过 [sysbench](https://www.percona.com/blog/tag/sysbench/) 等数据库压测工具进行压测来体验弹性伸缩功能。
+
+销毁环境
+
+```shell
+kubectl delete tidbcluster auto-scaling-demo -n <namespace>
+kubectl delete tidbmonitor auto-scaling-demo -n <namespace>
+kubectl delete tidbclusterautoscaler auto-scaling-demo -n <namespace>
```
Comment (DanielZhangQD) [inline zh/enable-tidb-cluster-auto-scaling.md]: Maybe `kubectl delete -f examples/auto-scale/  -n <namespace>`?

File: zh/enable-tidb-cluster-auto-scaling.md [hunk 11]
```diff
@@ -0,0 +1,144 @@
...
+```
+
+## 配置 TidbClusterAutoScaler
+
+对比起无状态的 Web 服务，一个分布式数据库软件对于实例的伸缩往往是非常敏感的。我们需要保证每次弹性伸缩之间存在一定的间隔，从而避免引起频繁的弹性伸缩。
```
Comment (DanielZhangQD) [inline zh/enable-tidb-cluster-auto-scaling.md]:
```suggestion
相比无状态的 Web 服务，一个分布式数据库软件对于实例的伸缩往往是非常敏感的。我们需要保证每次弹性伸缩之间存在一定的间隔，从而避免引起频繁的弹性伸缩。
```

File: zh/enable-tidb-cluster-auto-scaling.md (hunk 6 above)
Comment (DanielZhangQD) [reply zh/enable-tidb-cluster-auto-scaling.md]: The words in the description do not match with those in the below yamls.

File: zh/enable-tidb-cluster-auto-scaling.md [hunk 12]
```diff
@@ -0,0 +1,144 @@
...
+你可以通过 `spec.tikv.scaleOutThreshold` 和 `spec.tikv.scaleInThreshold` 来配置每两次弹性伸缩之间的时间间隔，对于 tidb 也同样如此。
+
+```yaml
```
Comment (DanielZhangQD) [inline zh/enable-tidb-cluster-auto-scaling.md]: There is not a full example of autoscaler after going through this doc, how could users know what can they configure for the autoscaler? @ran-huang What do you think?

Comment (Yisaer) [reply zh/enable-tidb-cluster-auto-scaling.md]: I think you should provide more detail about the what full example means.

File: zh/enable-tidb-cluster-auto-scaling.md (hunk 9 above)
Comment (Yisaer) [reply zh/enable-tidb-cluster-auto-scaling.md]: no need, updated.

File: zh/enable-tidb-cluster-auto-scaling.md (hunk 10 above)
Comment (Yisaer) [reply zh/enable-tidb-cluster-auto-scaling.md]: ditto.

File: zh/enable-tidb-cluster-auto-scaling.md [hunk 13]
```diff
@@ -0,0 +1,144 @@
...
+```
+
+开启 Operator Webhook 特性:
+
+```yaml
+admissionWebhook:
+  create: true
+  mutation:
+    pods: true
+```
```
[the remark this answers is not in the document]
Comment (Yisaer) [reply zh/enable-tidb-cluster-auto-scaling.md]: updated.

**Comment by Yisaer** [comment]
@DanielZhangQD @ran-huang PTAL

File: zh/enable-tidb-cluster-auto-scaling.md [hunk 14]
```diff
@@ -0,0 +1,146 @@
...
+开启 Operator Webhook 特性:
+
+    ```yaml
+    admissionWebhook:
+      create: true
+      mutation:
+        pods: true
+    ```
+
+2. 安装 / 更新 operator
```
Comment (ran-huang) [inline zh/enable-tidb-cluster-auto-scaling.md]: Remove extra spaces.
```suggestion
2. 安装/更新 operator
```

File: zh/enable-tidb-cluster-auto-scaling.md [hunk 15]
```diff
@@ -0,0 +1,146 @@
...
+
+    修改完 `values.yaml` 文件中上述配置项以后进行 TiDB-Operator 部署或者更新。安装与更新 Operator 请参考[在 Kubernetes 上部署 TiDB Operator](deploy-tidb-operator.md)
```
Comment (ran-huang) [inline zh/enable-tidb-cluster-auto-scaling.md]:
```suggestion
    修改完 `values.yaml` 文件中上述配置项以后进行 TiDB-Operator 部署或者更新。安装与更新 Operator 请参考[在 Kubernetes 上部署 TiDB Operator](deploy-tidb-operator.md)。
```

File: zh/enable-tidb-cluster-auto-scaling.md [hunk 16]
```diff
@@ -0,0 +1,146 @@
...
+
+## 了解 TidbClusterAutoScaler
+
+我们通过 `TidbClusterAutoScaler` CR 对象来控制 TiDB 集群的弹性伸缩行为，如果你曾经使用过 [Horizontal Pod Autoscaler](https://kubernetes.io/zh/docs/tasks/run-application/horizontal-pod-autoscale/)， 那么你一定会对这个概念感到非常熟悉。以下是一个 TiKV 的弹性伸缩例子。
```
Comment (ran-huang) [inline zh/enable-tidb-cluster-auto-scaling.md]:
```suggestion
我们通过 `TidbClusterAutoScaler` CR 对象来控制 TiDB 集群的弹性伸缩行为，如果你曾经使用过 [Horizontal Pod Autoscaler](https://kubernetes.io/zh/docs/tasks/run-application/horizontal-pod-autoscale/)，那么你一定会对这个概念感到非常熟悉。以下是一个 TiKV 的弹性伸缩例子。
```

File: zh/enable-tidb-cluster-auto-scaling.md [hunk 17]
```diff
@@ -0,0 +1,146 @@
...
+            resource:
+              name: "cpu"
+              target:
+                type: "Utilization"
+                averageUtilization: 80
+    ```
+
+对于 TiDB 组件，你可以通过 `spec.tidb` 来进行配置，目前 TiKV 与 TiDB 的弹性伸缩 API 相同。
+
+在 `TidbClusterAutoScaler` 对象中，`cluster` 属性代表了需要被弹性调度的 TiDB 集群，通过 name 与 namespace 来标识。 由于 `TidbClusterAutoScaler` 组件需要通过指标采集组件抓取相关资源使用情况，我们需要提供对应的指标采集与查询服务给 `TidbClusterAutoScaler`。`monitor` 属性则代表了与之相关连的 TidbMonitor 对象。 如果你不了解 TidbMonitor，可以参考[通过 TidbMonitor 监控 TiDB 集群](monitor-using-tidbmonitor.md)。
```
Comment (ran-huang) [inline zh/enable-tidb-cluster-auto-scaling.md]:
```suggestion
在 `TidbClusterAutoScaler` 对象中，`cluster` 属性代表了需要被弹性调度的 TiDB 集群，通过 name 与 namespace 来标识。由于 `TidbClusterAutoScaler` 组件需要通过指标采集组件抓取相关资源使用情况，我们需要提供对应的指标采集与查询服务给 `TidbClusterAutoScaler`。`monitor` 属性则代表了与之相关连的 TidbMonitor 对象。如果你不了解 TidbMonitor，可以参考[通过 TidbMonitor 监控 TiDB 集群](monitor-using-tidbmonitor.md)。
```

File: zh/enable-tidb-cluster-auto-scaling.md [hunk 18]
```diff
@@ -0,0 +1,193 @@
...
+kubectl delete tidbmonitor auto-scaling-demo -n <namespace>
+kubectl delete tidbclusterautoscaler auto-scaling-demo -n <namespace>
+```
+
+## 配置 TidbClusterAutoScaler
+
+1. 设置弹性伸缩间隔
+
+相比无状态的 Web 服务，一个分布式数据库软件对于实例的伸缩往往是非常敏感的。我们需要保证每次弹性伸缩之间存在一定的间隔，从而避免引起频繁的弹性伸缩。
+你可以通过 `spec.tikv.scaleOutThreshold` 和 `spec.tikv.scaleInThreshold` 来配置每两次弹性伸缩之间的时间间隔(秒)，对于 TiDB 也同样如此。
```
Comment (DanielZhangQD) [inline zh/enable-tidb-cluster-auto-scaling.md]: The below example does not include this two fields.

File: zh/enable-tidb-cluster-auto-scaling.md (hunk 12 above)
Comment (DanielZhangQD) [reply zh/enable-tidb-cluster-auto-scaling.md]: For a full example, I mean that users may want to know all of the fields and descriptions that can be configured in the autoscaler.

File: zh/enable-tidb-cluster-auto-scaling.md (hunk 18 above)
Comment (Yisaer) [reply zh/enable-tidb-cluster-auto-scaling.md]: updated.

**Comment by Yisaer** [comment]
@DanielZhangQD @ran-huang  PTAL

File: zh/enable-tidb-cluster-auto-scaling.md [hunk 19]
```diff
@@ -0,0 +1,193 @@
...
+        metrics:
+          - type: "Resource"
+            resource:
+              name: "cpu"
+              target:
+                type: "Utilization"
+                averageUtilization: 80
+    ```
+
+4. 配置指标时间窗口
```
Comment (DanielZhangQD) [inline zh/enable-tidb-cluster-auto-scaling.md]: Do we need to add a description for the below fields?
```
        ScaleOutThreshold *int32 `json:"scaleOutThreshold,omitempty"`

	// ScaleInThreshold describe the consecutive threshold for the auto-scaling,
	// if the consecutive counts of the scale-in result in auto-scaling reach this number,
	// the auto-scaling would be performed.
	// If not set, the default value is 5.
	// +optional
	ScaleInThreshold *int32 `json:"scaleInThreshold,omitempty"`
```

File: zh/enable-tidb-cluster-auto-scaling.md [hunk 20]
```diff
@@ -0,0 +1,193 @@
...
+开启 Operator Webhook 特性:
+
+    ```yaml
+    admissionWebhook:
+      create: true
+      mutation:
+        pods: true
+    ```
+
+2. 安装/更新 operator
```
Comment (DanielZhangQD) [inline zh/enable-tidb-cluster-auto-scaling.md]:
```suggestion
2. 安装/更新 TiDB Operator
```

File: zh/enable-tidb-cluster-auto-scaling.md [hunk 21]
```diff
@@ -0,0 +1,193 @@
...
+
+    修改完 `values.yaml` 文件中上述配置项以后进行 TiDB-Operator 部署或者更新。安装与更新 Operator 请参考[在 Kubernetes 上部署 TiDB Operator](deploy-tidb-operator.md)。
```
Comment (ran-huang) [inline zh/enable-tidb-cluster-auto-scaling.md]:
```suggestion
    修改完 `values.yaml` 文件中上述配置项以后进行 TiDB Operator 部署或者更新。安装与更新 Operator 请参考[在 Kubernetes 上部署 TiDB Operator](deploy-tidb-operator.md)。
```

File: zh/enable-tidb-cluster-auto-scaling.md [hunk 22]
```diff
@@ -0,0 +1,193 @@
...
+kubectl delete tidbcluster auto-scaling-demo -n <namespace>
+kubectl delete tidbmonitor auto-scaling-demo -n <namespace>
+kubectl delete tidbclusterautoscaler auto-scaling-demo -n <namespace>
+```
+
+## 配置 TidbClusterAutoScaler
+
+1. 设置弹性伸缩间隔
+
+相比无状态的 Web 服务，一个分布式数据库软件对于实例的伸缩往往是非常敏感的。我们需要保证每次弹性伸缩之间存在一定的间隔，从而避免引起频繁的弹性伸缩。
```
Comment (DanielZhangQD) [inline zh/enable-tidb-cluster-auto-scaling.md]: The sentences in the list should be indented, right? @ran-huang

File: zh/enable-tidb-cluster-auto-scaling.md [hunk 23]
```diff
@@ -0,0 +1,193 @@
...
+
+2. 安装/更新 TiDB Operator
+
+    修改完 `values.yaml` 文件中上述配置项以后进行 TiDB Operator 部署或者更新。安装与更新 Operator 请参考[在 Kubernetes 上部署 TiDB Operator](deploy-tidb-operator.md)。
+
+## 了解 TidbClusterAutoScaler
+
+我们通过 `TidbClusterAutoScaler` CR 对象来控制 TiDB 集群的弹性伸缩行为，如果你曾经使用过 [Horizontal Pod Autoscaler](https://kubernetes.io/zh/docs/tasks/run-application/horizontal-pod-autoscale/)，那么你一定会对这个概念感到非常熟悉。以下是一个 TiKV 的弹性伸缩例子。
+
+    ```yaml
```
Comment (DanielZhangQD) [inline zh/enable-tidb-cluster-auto-scaling.md]: No need to indent the code block here, right? @ran-huang

File: zh/enable-tidb-cluster-auto-scaling.md (hunk 22 above)
Comment (ran-huang) [reply zh/enable-tidb-cluster-auto-scaling.md]: Yes, you are right.

File: zh/enable-tidb-cluster-auto-scaling.md (hunk 23 above)
Comment (ran-huang) [reply zh/enable-tidb-cluster-auto-scaling.md]: Yes. If the code block is part of the text, no need to indent. Only indent code blocks when they are under a list.

Comment (Yisaer) [reply zh/enable-tidb-cluster-auto-scaling.md]: updated

File: zh/enable-tidb-cluster-auto-scaling.md (hunk 22 above)
Comment (Yisaer) [reply zh/enable-tidb-cluster-auto-scaling.md]: updated

**Pull request merged**